### ISSUE TITLE
nixos/lib/make-disk-image: performance regression since 16.03

### ISSUE THREAD

**Issue by domenkozar** [NixOS/nixpkgs#20471]
24f8cf08 changes `nixos/lib/make-disk-image` to use nixos-install, but that doesn't copy the closure correctly so it takes way longer to build.

To reproduce run:

```
         nix-build \
             '<nixpkgs/nixos>' \
             -A config.system.build.amazonImage \
             --arg configuration "{ imports = [ <nixpkgs/nixos/maintainers/scripts/ec2/amazon-image.nix> ]; ec2.hvm = true; }"
```

And to make things worse, `nixos-install` uses `-j 1` by default. This is considered a regression, ami image takes now ~35min to build.

cc @obadz

**Comment by obadz** [comment]
What do you mean by 'downloads files from the internet' ? Shouldn't that be impossible from a chrooted build?

Should `-j 1` matter given that everything should be in the cache provided from inside the qemu VM so nothing should be built?

I just ran the exact command you provided and it took 3m22s (I'm on c859ea0).

**Comment by domenkozar** [comment]
You don't get nixos-install fetching lots of files like:

```
*** Copying ‘/nix/store/n2mzj8h0rjp39di41jf762nzjhq6r4w2-libunwind-1.1’ from ‘/tmp/root/nix/store/n2mzj8h0rjp39d
i41jf762nzjhq6r4w2-libunwind-1.1’

fetching path ‘/nix/store/kkqf5mx287fxlsvxm2g8ys0wkfrbwgs1-libxcb-1.12’...

```

I'm on master branch.

**Comment by obadz** [comment]
I do, but that's just the nix-env inside the chroot inside nixos-install that's copying derivations from the nix store of the qemu VM onto the new filesystem that's being created.

Fetching here probably mean some substituter is involved, not the network?

**Comment by domenkozar** [comment]
Ah I see now! Didn't even realise it's not coming from http.

**Comment by obadz** [comment]
There is however an inefficiency with:

```
fixing missing hash on ‘/nix/store/8jwqlpyxp5c2rl5g7yspdfbh58dsciwx-xz-5.2.2’
updating size field on ‘/nix/store/8jwqlpyxp5c2rl5g7yspdfbh58dsciwx-xz-5.2.2’ to 328320
```

Essentially we don't yet have the ability to create a valid nix store inside a qemu VM that contains the transitive closure of some derivation on the mother system. The thing will boot and be normally usable, but nix operations will not recognize these packages without this step:

https://github.com/NixOS/nixpkgs/blob/master/nixos/lib/make-disk-image.nix#L74-L76

And that step takes time…

Still doesn't explain how you get to 35 minutes. How long does it take you to build it the second time around after a trivial config change?

**Comment by domenkozar** [comment]
@obadz this is on EC2. I'll test with 16.09 and 16.03 and compare timings.

I agree that it would be good to get size information etc into exported graph so we don't have to do so much IO. Does `nix-store --load-db` support such format in the first place?

**Comment by domenkozar** [comment]
@obadz opened https://github.com/NixOS/nix/issues/1134

**Comment by domenkozar** [comment]
The performance regression is real. This is on EC2:

16.03:

```
1.33user 0.75system 8:14.34elapsed 0%CPU (0avgtext+0avgdata 252760maxresident)k
0inputs+0outputs (0major+67380minor)pagefaults 0swaps
```

16.09:

```
/nix/store/sjlfm79f466ywia405615c2pmafraimp-nixos-disk-image
1.69user 1.11system 31:05.59elapsed 0%CPU (0avgtext+0avgdata 291044maxresident)k
0inputs+0outputs (0major+77253minor)pagefaults 0swaps
```

**Comment by domenkozar** [comment]
Sadly we don't have any jobs on hydra to track build times of image builds

**Comment by domenkozar** [comment]
cc @edolstra @rbvermaa

**Comment by domenkozar** [comment]
@obadz maybe we could still `cp` the closure and then use nixos-install?

**Comment by copumpkin** [comment]
@domenkozar could it be that this is now doing a lot more in qemu on EC2? Keep in mind that EC2 doesn't support nested virtualization, so everything that happens inside the VM is fully emulated and obscenely slow. That's why I keep pushing for ways to build images that don't involve any VM, because everything I do is in EC2 😄

**Comment by obadz** [comment]
That would explain the massive perf differential between local and EC2. I vote for trying to fix https://github.com/NixOS/nix/issues/1134 before considering other avenues.

**Comment by copumpkin** [comment]
Well, I still think we should stop using VMs to build images, as much as possible, because it's still far slower than it needs to be. But yes, that's also a problem.

**Comment by obadz** [comment]
We'd need an unprivileged mechanism to write an ext2 filesystem (and a few more things). Can we get fuse-ext2 to work inside the nix sandbox? Without it I think it will be hard.

**Comment by copumpkin** [comment]
There's already code that does that with debugfs in nixpkgs today. Not sure how fast it is. There's also https://github.com/rumpkernel/fs-utils and a similar project for Linux that I can't remember right now. I don't think fuse will really work for us.

**Comment by obadz** [comment]
It might also be btw that the hashing required to validate the nix-store (which we ought to be able to get rid off) is much much slower than the IO required to do the copying, in which case qemu will become much less of a burden once we fix NixOS/nix#1134

**Comment by domenkozar** [comment]
https://github.com/rumpkernel/wiki/wiki/Repo:-fs-utils does looks useful here.

**Comment by domenkozar** [comment]
@obadz out of 60min in total, 13min is spend rewriting hashes while 40min is spent in nixos-install.

**Comment by copumpkin** [comment]
@domenkozar https://github.com/lkl/linux looks like it's planning to be something similar to the rump project for linux, so it might be possible to get something like `fs-utils` against all the filesystem modules we normally use and love 😄

**Comment by obadz** [comment]
Hmmm I fail to understand why `nixos-install` takes any more time than https://github.com/NixOS/nixpkgs/commit/24f8cf08cc897f0b3c2ff4118fe30c38eed85aa6#diff-7fce4d49800213db1196b42209660eb4L94 (line 94-95) as that's essentially what nixos-install runs…

Are you sure the performance is restored when you revert that commit?

**Comment by obadz** [comment]
Nevermind, I take that back. Seems like the old way copied the store path directly inside the target rather than first installing it inside the qemu VM and then using substituters.

**Comment by domenkozar** [comment]
@copumpkin I looked into fs-utils and we won't be able to run `nixos-install` with it since there's no way to chroot. Did I miss something?

**Comment by domenkozar** [comment]
It might be worth trying out methods in https://wiki.debian.org/ManipulatingISOs#Loopmount_an_ISO_Without_Administrative_Privileges

**Comment by cleverca22** [comment]
at a glance, i think those are just a group of setuid root tools that do things for you and enforce some security rules

**Comment by copumpkin** [comment]
@domenkozar fundamentally, it doesn't seem like `nixos-install` should need `chroot`, right?

**Comment by copumpkin** [comment]
Trying to build a NixOS image inside EC2 and it's miserable. We need to do something about this. Going to start poking around at minimizing (probably not to 0) the amount of VM time we need for it...

**Comment by copumpkin** [comment]
@cleverca22 @domenkozar @obadz see #21943 for current WIP

**Comment by Nadrieril** [comment]
Since https://github.com/NixOS/nix/issues/1134 has been closed, is there now a way to avoid doing `nix-store --verify --check-contents` ? I can't seem to find how to use the json exported data from `exportReferencesGraph` anywhere.

**Comment by domenkozar** [comment]
That feature will land in Nix 1.12 so until it's released we can't use it :)

**Comment by Nadrieril** [comment]
Ok, thx :)

**Comment by volth** [comment]
```exportReferencesGraph``` and ```pathsFromGraph``` could be replaced with something like:

```
      # use instead of "perl ${pkgs.pathsFromGraph} /tmp/xchg/closure", no difference
      storePaths = toplevel:
        hostpkgs.runCommand "storepaths" { preferLocalBuild = true; } ''
          ${hostpkgs.nix}/bin/nix-store -qR ${toplevel} > $out
        '';

      # use instead of "printRegistration=1 perl ${pkgs.pathsFromGraph} /tmp/xchg/closure", it has valid sizes and hashes
      dumpDb = toplevel:
        hostpkgs.runCommand "dumpdb" {
          preferLocalBuild = true;
          perlcode = hostpkgs.writeScript "xxx.pl" ''
            my %storePaths={};
            open SP, "<${storePaths toplevel}";
            while (<SP>) {
              $storePaths{$_} = 1;
            }
            close SP;

            open DB, '-|', '${hostpkgs.nix}/bin/nix-store --dump-db' or die $!;
            while (<DB>) {
              my $path  = $_;   print $path  if $storePaths{$path};
              my $hash  = <DB>; print $hash  if $storePaths{$path};
              my $size  = <DB>; print $size  if $storePaths{$path};
                          <DB>; print "\n"   if $storePaths{$path};
              my $count = <DB>; print $count if $storePaths{$path};
              chomp $count;
              for (my $i = 0; $i < $count; ++$i) {
                my $dep = <DB>; print $dep   if $storePaths{$path};
              }
            }
            close DB;
          '';
        } ''
          ${hostpkgs.perl}/bin/perl $perlcode > $out
        '';
```

**Comment by copumpkin** [comment]
@volth your version only works if sandboxing is turned off :/

**Comment by volth** [comment]
@copumpkin that's bad. But NixOS/nix#1134 fix is not in ```17.03``` yet (even as ```nixUnstable```) and recalculating the hashes from the content (besides being slow) could legalize nixstore corruption or bitrot on the target.

**Comment by domenkozar** [comment]
FYI: this is being addressed in https://github.com/NixOS/nixpkgs/pull/24964

**Comment by copumpkin** [comment]
Strictly speaking, my PR isn't the only (or smallest) way to fix it, but it does give us some other advantages while also fixing this 😁

**Comment by fpletz** [comment]
Fixed in master by #24964.
